### PR TITLE
TINKERPOP-3028 Translators using grammar

### PR DESCRIPTION
**Pull request by spmallette**
https://issues.apache.org/jira/browse/TINKERPOP-3028

Added a normalizer for language and an anonymizer implementation as well. Extract parameter names. Updated test setups to use the new translators.

Worth noting that this isn't quite complete - see all the linked JIRA on TINKERPOP-3028. All of that will need to be done still but this seems like a good enough chunk to merge now and in its current form may allow some other work to proceed more easily.

VOTE +1

### CODE REVIEW

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/grammar/TraversalMethodVisitor.java
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/grammar/TraversalMethodVisitor.java:1]: `visitTraversalMethod_option_Merge_Map_Cardinality` have middle argument (Map) defined in grammar as `genericLiteralMapNullableArgument`, but we handle only  `genericLiteralMap` and `nullLiteral` cases, not `variable`.

   ```
 @Override
    public Traversal visitTraversalMethod_option_Merge_Map_Cardinality(final GremlinParser.TraversalMethod_option_Merge_Map_CardinalityContext ctx) {
        if (ctx.genericLiteralMapNullableArgument().nullLiteral() != null) {
            return this.graphTraversal.option(antlr.argumentVisitor.parseMerge(ctx.traversalMergeArgument()), (Map) null);
        }

        // proposed addition
        if (ctx.genericLiteralMapNullableArgument().variable() != null) {
            return graphTraversal.option(antlr.argumentVisitor.parseMerge(ctx.traversalMergeArgument()),
                    (Map) antlr.argumentVisitor.visitVariable(ctx.genericLiteralMapNullableArgument().variable()),
                    TraversalEnumParser.parseTraversalEnumFromContext(Cardinality.class, ctx.traversalCardinality()));
        }
        // end of proposed addition

        return graphTraversal.option(antlr.argumentVisitor.parseMerge(ctx.traversalMergeArgument()),
                (Map) new GenericLiteralVisitor(antlr).visitGenericLiteralMap(ctx.genericLiteralMapNullableArgument().genericLiteralMap()),
                TraversalEnumParser.parseTraversalEnumFromContext(Cardinality.class, ctx.traversalCardinality()));
    }
```

File: gremlin-dotnet/build/generate.groovy [hunk 1]
```diff
@@ -175,6 +137,11 @@ radishGremlinFile.withWriter('UTF-8') { Writer writer ->
             '            list.RemoveAt(0);\n' +
             '            return f.Invoke(g, parameters);\n' +
             '        }\n' +
+            '\n' +
+            '        public static bool HasTraversal(string scenarioName)\n' +
```
Comment (vkagamlyk) [inline gremlin-dotnet/build/generate.groovy:141]: where is `HasTraversal` used?

File: gremlin-python/src/main/python/gremlin_python/process/graph_traversal.py [hunk 2]
```diff
@@ -1424,6 +1428,10 @@ def ltrim(cls, *args):
     def l_trim(cls, *args):
         return cls.graph_traversal(None, None, Bytecode()).l_trim(*args)
 
+    @classmethod
```
Comment (vkagamlyk) [inline gremlin-python/src/main/python/gremlin_python/process/graph_traversal.py:1431]: Looks like there's an issue with merge.
Duplicate `l_trim` and `r_trim` function definition (4 times in total).

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/AnonymizedTranslatorVisitor.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    @Override
+    public Void visitFloatLiteral(final GremlinParser.FloatLiteralContext ctx) {
+        final String floatLiteral = ctx.getText().toLowerCase();
+
+        // check suffix
+        final int lastCharIndex = floatLiteral.length() - 1;
+        final char lastCharacter = floatLiteral.charAt(lastCharIndex);
+        switch (lastCharacter) {
```
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/AnonymizedTranslatorVisitor.java:140]: do we need special handling for special values like NaN and Infinity?

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/GroovyTranslateVisitor.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            case 's':
+                // parse S/s as short
+                sb.append("new Short(");
+                sb.append(integerLiteral, 0, lastCharIndex);
+                sb.append(")");
+                break;
+            case 'i':
+            case 'l':
+                // parse I/i and L/l as Integer and Long respectively
+                sb.append(integerLiteral, 0, lastCharIndex).append(lastCharacter);
```
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/GroovyTranslateVisitor.java:78]: nit: no need special handling for i/l case, it's same as default

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/JavaTranslateVisitor.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        super("g");
+    }
+
+    public JavaTranslateVisitor(final String graphTraversalSourceName) {
+        super(graphTraversalSourceName);
+    }
+
+    @Override
+    public Void visitStructureVertex(final GremlinParser.StructureVertexContext ctx) {
+        sb.append("new DetachedVertex(");
```
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/JavaTranslateVisitor.java:56]: [Discuss] what is better to use in this case 'DetachedVertex' or `ReferenceVertex`?

File: gremlin-dotnet/build/generate.groovy (hunk 1 above)
Comment (spmallette) [reply gremlin-dotnet/build/generate.groovy:141]: it was being used, but i switched to statically defining the unsupported lambda tests. removing it.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/GoTranslateVisitor.java [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        final String dtString = ctx.getChild(2).getText();
+        final Date dt = DatetimeHelper.parse(removeFirstAndLastCharacters(dtString));
+        sb.append("time.UnixMilli(" + dt.getTime() + ")");
+        return null;
+    }
+
+    @Override
+    public Void visitInfLiteral(final GremlinParser.InfLiteralContext ctx) {
+        if (ctx.SignedInfLiteral().getText().equals("-Infinity"))
+            sb.append("math.Inf(-11)");
```
[the remark this answers is not in the document]
Comment (spmallette) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/GoTranslateVisitor.java:58]: oops - i even wrote the test incorrectly. good find.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/JavaTranslateVisitor.java (hunk 5 above)
Comment (spmallette) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/JavaTranslateVisitor.java:56]: I suppose `ReferenceVertex` is better. This is truly a "reference" I guess.

File: gremlin-python/src/main/python/gremlin_python/process/graph_traversal.py (hunk 2 above)
Comment (spmallette) [reply gremlin-python/src/main/python/gremlin_python/process/graph_traversal.py:1431]: thanks for noticing that.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/grammar/TraversalMethodVisitor.java
Comment (spmallette) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/grammar/TraversalMethodVisitor.java:1]: adding it.

File: gremlin-python/src/main/python/gremlin_python/process/graph_traversal.py [hunk 7]
```diff
@@ -2112,6 +2116,9 @@ def l_trim(*args):
     return __.l_trim(*args)
 
 
+def l_trim(*args):
```
Comment (vkagamlyk) [inline gremlin-python/src/main/python/gremlin_python/process/graph_traversal.py:2119]: one more duplicate `l_trim` definition

File: gremlin-python/src/main/python/gremlin_python/process/graph_traversal.py [hunk 8]
```diff
@@ -2236,6 +2243,10 @@ def r_trim(*args):
     return __.r_trim(*args)
 
 
+def r_trim(*args):
```
Comment (vkagamlyk) [inline gremlin-python/src/main/python/gremlin_python/process/graph_traversal.py:2246]: and one more duplicate `r_trim` definition

File: docs/src/upgrade/release-4.x.x.asciidoc [hunk 9]
```diff
@@ -118,11 +118,41 @@ passes lists with no elements matching the predicate.
...
+language to translate to. Consider the following example:
+
+[source,text]
+----
+gremlin> GremlinTranslator.translate("g.V().out('knows')", Translator.GO)
+==>g.V().Out("knows")
+----
+
+Further note that Gremlin language variants produce `gremlin-language` compliant strings directly since bytecode was
+removed. As a result, all translators in .NET, Python, C# and Javascript have been removed.
```
Comment (Cole-Greer) [inline docs/src/upgrade/release-4.x.x.asciidoc:153]:
```suggestion
removed. As a result, all translators in .NET, Python, Go and Javascript have been removed.
```

**Comment by Cole-Greer** [comment]
Thanks for all of the work on this, it is a big step forward to have modern gremlin-lang based translators.

I primarily focused my review on the Go and JS translators here as it is a big PR and hard to capture everything at once. Prior to the release of TP4, I would like to add some tests to ensure successful translation of some of the more obscure literals, however I think that can be out of scope for this PR as the new translators are at least as well tested as the old ones.

VOTE +1

**Pull request merged**